tests/core/test_utilities.py

```diff
@@ -913,6 +914,81 @@ def test_fit_plane_to_points():
     )
 
 
+DEFAULT_PRINCIPAL_AXES = [[0.0, 0.0, 1.0], [0.0, 1.0, 0.0], [-1.0, 0.0, 0.0]]
+
+CASE_RANK0 = (
+    0,
+    [[0, 0, 0], [0, 0, 0]],
+    [DEFAULT_PRINCIPAL_AXES],
+)
+CASE_RANK1 = (
+    1,
+    [[0, 0, 0], [1, 0, 0]],
+    [[1.0, 0.0, 0.0], [0.0, 0.0, 1.0], [0.0, -1.0, 0.0]],
+)
+
+CASE_RANK2 = (
+    2,
+    [[1, 2, 3], [3, 2, 1]],
+    [
+        [-0.70710678, 0.0, 0.70710678],
+        [0.0, 1.0, 0.0],
+        [-0.70710678, 0.0, -0.70710678],
+    ],
+)
+CASE_RANK3 = (
+    3,
+    pv.examples.load_airplane().points,
+    [
+        [8.1159601e-07, -9.9255711e-01, -1.2178054e-01],
+        [-1.0000000e00, -8.0911440e-07, -6.9828459e-08],
+        [-2.9225653e-08, 1.2178054e-01, -9.9255711e-01],
+    ],
+)
+
+
+@pytest.mark.parametrize(
+    ('rank', 'points', 'expected_axes'),
+    [CASE_RANK0, CASE_RANK1, CASE_RANK2, CASE_RANK3],
+    ids=['rank0', 'rank1', 'rank2', 'rank3'],
+)
+def test_principal_axes(rank, points, expected_axes):
+    assert np.linalg.matrix_rank(points) == rank
+
+    axes = principal_axes(points)
+    assert np.allclose(axes, expected_axes, atol=1e-7)
+
+    assert np.array_equal(np.cross(axes[0], axes[1]), axes[2])
+    assert np.allclose(np.linalg.norm(axes[0]), 1)
+    assert np.allclose(np.linalg.norm(axes[1]), 1)
+    assert np.allclose(np.linalg.norm(axes[2]), 1)
+
+    assert type(axes) is np.ndarray
```

Comment:
Are you doing exact type checks on purpose? Or can we use
```suggestion
    assert isinstance(axes, np.ndarray)
```
? I see that `np.linalg.eigh()` nicely returns a `pyvista_ndarray()` for input of the same type, so I assume this is on purpose, but then a comment might help future readers.

Reply:
We can keep the default behaviour and keep the `T` -> `T` type mapping  (and thus return `pyvista_ndarray` if we have `pyvista_ndarray` in.

Reply:
> We can keep the default behaviour and keep the `T` -> `T` type mapping (and thus return `pyvista_ndarray` if we have `pyvista_ndarray` in.

But returning `ndarray` for `list` or other compatible sequence input might throw a wrench in that...

Reply:
Ah yes, I guess I meant we can always return an instance of `np.ndarray`, and therefore allow for things like `pyvista_ndarray` -> `pyvista_ndarray`. Basically just remove any type casting from the output.

Reply:
This has been updated to remove any type casting.